optimize chunk map

hayanesuru opened the pull request:
1. revert volatile in `LevelChunkSection#nonEmptyBlockCount`
2. add non-sync map of `ChunkHolderManager#chunkHolders` for main thread
3. add non-sync map of `ServerChunkCache#fullChunks` for main thread
4. remove unused `SimpleBitStorage#divideMulUnsigned` `SimpleBitStorage#divideAddUnsigned` to optimize object layout (after: Hotspot >=15 64-bit == 64, COOPS CCPS == 56, CCPS == 64)
5. remove type cast in `PalettedContainer#readPalette(PalettedContainer$Data, int)`
6. opt `LevelChunk#getBlockStateFinal(int, int, int)` inline min section
7. weaken the memory barrier (full -> acquire) in `PalettedContainer#get(int)`
8. opt `Level#getBlockState(BlockPos)`
9. remove tagArray in `Optimize isEyeInFluid`

```text
2	0	net/minecraft/server/level/ServerLevel.java
5	5	net/minecraft/util/SimpleBitStorage.java
175	120	net/minecraft/world/entity/Entity.java
7	0	net/minecraft/world/entity/ExperienceOrb.java
21	6	net/minecraft/world/entity/InsideBlockEffectApplier.java
30	21	net/minecraft/world/entity/LivingEntity.java
8	1	net/minecraft/world/entity/item/ItemEntity.java
75	0	net/minecraft/world/level/BlockGetter.java
17	4	net/minecraft/world/level/Level.java
1	1	net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
7	0	net/minecraft/world/level/block/state/BlockBehaviour.java
26	5	net/minecraft/world/level/chunk/LevelChunk.java
38	1	net/minecraft/world/level/chunk/LevelChunkSection.java
14	2	net/minecraft/world/level/chunk/PalettedContainer.java
10	1	net/minecraft/world/level/material/FluidState.java
 15 files changed, 436 insertions(+), 167 deletions(-)
```

note: Plain reads are consistently faster than atomic (eg. memory_order_seq_cst / volatile), but the difference often becomes negligible compared to cache miss costs. cache efficiency often yields larger gains than micro-optimizing synchronization primitives

MartijnMuijsers commented on leaf-server/minecraft-patches/features/0251-Cache-block-path-type.patch:
```diff
@@ -36,24 +36,30 @@ index 57aad048034005543a72556e990b53db8deebfee..6e6ee6f25f7cab0a6235d9e583ed5c01
++            tagFlag |= this.asState().is(net.minecraft.tags.BlockTags.FENCES) ? (byte) 2 : (byte) 0;
++            tagFlag |= this.asState().is(net.minecraft.tags.BlockTags.CLIMBABLE) ? (byte) 4 : (byte) 0;
++            tagFlag |= this.getBlock() instanceof net.minecraft.world.level.block.FenceGateBlock ? (byte) 8 : (byte) 0;
++            tagFlag |= this.getBlock() instanceof net.minecraft.world.level.block.TrapDoorBlock ? (byte) 16 : (byte) 0;
```
Obviously faster to calculate the whole tagFlag in one go, and do it all using int, and then cast to byte only once in the end.

MartijnMuijsers commented on leaf-server/minecraft-patches/features/0285-optimize-Entity-getOnPos.patch:
```diff
@@ -0,0 +1,52 @@
++                    int floor2 = Mth.floor(this.position.z);
++                    return new BlockPos(floor, floor1, floor2);
++                }
++                return (!(yOffset <= 0.5) || ((blockState.tagFlag & 2) != 2))
```
Well teeeeeeeeeeechnically, if you know the result is "something or 0" it's usually better to check != 0 or == 0, because then the machine code will check the zero flag (`jz` instruction) instead of doing an == comparison, which is potentially faster.

MartijnMuijsers replied:
(JIT will usually do this for you, but potentially only C2)

hayanesuru committed:
rebase

hayanesuru committed:
optimize LivingEntity#travel

hayanesuru committed:
cleanup

hayanesuru committed:
Replace fluid height map

hayanesuru committed:
reuse array list in Entity#collide

hayanesuru committed:
Merge branch 'ver/1.21.8' into opt/chunk-map

hayanesuru committed:
cleanup

hayanesuru committed:
fix fire and liquid collision shape

hayanesuru committed:
fix checkInside

hayanesuru committed:
inline betweenClosed

hayanesuru committed:
cleanup

hayanesuru committed:
optimize getOnPos

hayanesuru committed:
optimize equals in getOnPos

hayanesuru committed:
update mainSupportingBlockPos on dirty

hayanesuru committed:
cleanup

hayanesuru committed:
rename

hayanesuru committed:
Merge branch 'ver/1.21.8' into opt/chunk-map

# Conflicts:
#	leaf-server/minecraft-patches/features/0207-SparklyPaper-Parallel-world-ticking.patch
#	leaf-server/minecraft-patches/features/0223-Prevent-double-chunk-retrieving-in-entity-fluid-push.patch
#	leaf-server/minecraft-patches/features/0240-Optimise-player-movement-checks.patch
#	leaf-server/minecraft-patches/features/0247-Optimize-isEyeInFluid.patch
#	leaf-server/minecraft-patches/features/0265-optimize-waypoint.patch
#	leaf-server/minecraft-patches/features/0266-Paw-optimization.patch
#	leaf-server/minecraft-patches/features/0272-optimize-checkInsideBlocks-calls.patch

hayanesuru committed:
Merge branch 'ver/1.21.8' into opt/chunk-map

hayanesuru committed:
merge same patch

hayanesuru committed:
Merge branch 'ver/1.21.8' into opt/chunk-map

hayanesuru committed:
Merge branch 'ver/1.21.8' into opt/chunk-map

Taiyou06 committed:
Merge remote-tracking branch 'origin/ver/1.21.8' into opt/chunk-map

Taiyou06 committed:
rebase and remove properly

hayanesuru committed:
[ci skip] cleanup

hayanesuru committed:
[ci skip] Merge branch 'ver/1.21.8' into opt/chunk-map

# Conflicts:
#	leaf-server/minecraft-patches/features/0245-Optimize-isEyeInFluid.patch
#	leaf-server/minecraft-patches/features/0263-optimize-waypoint.patch
#	leaf-server/minecraft-patches/features/0264-Paw-optimization.patch
#	leaf-server/minecraft-patches/features/0270-optimize-checkInsideBlocks-calls.patch

Taiyou06 committed:
Merge remote-tracking branch 'origin/ver/1.21.8' into opt/chunk-map

Taiyou06 committed:
rebase and rebuild

hayanesuru committed:
fix async locator

hayanesuru committed:
remove async locator

hayanesuru committed:
cleanup

hayanesuru committed:
Merge branch 'ver/1.21.8' into opt/chunk-map

Taiyou06 committed:
Merge remote-tracking branch 'origin/ver/1.21.8' into opt/chunk-map

Taiyou06 committed:
rebase
